base/multidimensional.jl

```diff
@@ -127,7 +127,14 @@ module IteratorsMD
     @inline isless(I1::CartesianIndex{N}, I2::CartesianIndex{N}) where {N} = _isless(0, I1.I, I2.I)
     @inline function _isless(ret, I1::NTuple{N,Int}, I2::NTuple{N,Int}) where N
         newret = ifelse(ret==0, icmp(I1[N], I2[N]), ret)
-        _isless(newret, Base.front(I1), Base.front(I2))
+        f1, f2 = Base.front(I1), Base.front(I2)
+        return if isa(f1, Tuple{})
+            @assert isa(f2, Tuple{})
+            _isless(newret, f1, f2)
```

Comment:
Is `@assert` the right tool for these kind of hints? We have been talking about eliminating them unless in debug mode, so perhaps something like

```suggestion
            _isless(newret, f1, f2::Tuple{})
```
would be better?

Reply:
Thanks for your comment.
My original motivation to use `@assert` instead of `typeassert` was that otherwise JET will report false positive errors on the `typeassert`ions for some complicated cases where inference is unsure about the type of one tuple but sure about that of the other, e.g.:

<details><summary>example false positive reports</summary>

> under the definition of `_isless` below
> ```julia
>     @inline function _isless(ret, I1::NTuple{N,Int}, I2::NTuple{N,Int}) where N
>         newret = ifelse(ret==0, icmp(I1[N], I2[N]), ret)
>         f1, f2 = Base.front(I1), Base.front(I2)
>         if isa(f1, Tuple{})
>             return _isless(newret, f1, f2::Tuple{}) # L133
>         else
>             return _isless(newret, f1, f2::Tuple{Int,Vararg{Int}}) # L135
>         end
>     end
> ```
```julia
julia> report_call((Int, Tuple{Vararg{Int}}, Tuple{Int,Int,Vararg{Int}})) do ret, t1, t2
           Base.IteratorsMD._isless(ret, t1, t2)
       end
═════ 2 possible errors found ═════
┌ @ none:2 Base.getproperty(Base.IteratorsMD, :_isless)(ret, t1, t2)
│┌ @ /Users/aviatesk/julia/julia/base/multidimensional.jl:133 Core.typeassert(f2, Core.apply_type(Base.IteratorsMD.Tuple))
││ invalid builtin function call: Core.typeassert(f2::Union{Tuple{Int64}, Tuple{Int64, Int64, Vararg{Int64}}}, Core.apply_type(Base.IteratorsMD.Tuple)::Type{Tuple{}})
│└────────────────────────────────────────────────────────────
│┌ @ /Users/aviatesk/julia/julia/base/multidimensional.jl:135 Base.IteratorsMD._isless(newret, f1, Core.typeassert(f2, Core.apply_type(Base.IteratorsMD.Tuple, Base.IteratorsMD.Int, Core.apply_type(Base.IteratorsMD.Vararg, Base.IteratorsMD.Int))))
││┌ @ /Users/aviatesk/julia/julia/base/multidimensional.jl:135 Core.typeassert(f2, Core.apply_type(Base.IteratorsMD.Tuple, Base.IteratorsMD.Int, Core.apply_type(Base.IteratorsMD.Vararg, Base.IteratorsMD.Int)))
│││ invalid builtin function call: Core.typeassert(f2::Tuple{}, Core.apply_type(Base.IteratorsMD.Tuple, Base.IteratorsMD.Int, Core.apply_type(Base.IteratorsMD.Vararg, Base.IteratorsMD.Int)::Core.TypeofVararg)::Type{Tuple{Int64, Vararg{Int64}}})
││└────────────────────────────────────────────────────────────
```

</details>

and with `@assert` JET can forcibly exclude those cases (but yes, this is just because JET's bug report criteria is somewhat  looser on manually thrown exceptions than type failures...)

---

Anyway, I totally agree with `@assert` is very unnatural to be here, and there is no strong reason to make changes just for the sake of JET.
In the latest commit (85394123a6bae9c376d408ab758542ee63fe5cee) I focused on eliminating dynamic dispatches by adding type annotations (and also some possibilities for JET to report false positives).
Hopefully in this direction we can agree with taking these changes in ?

Reply:
Yes, your version looks reasonable to me.